### PR TITLE
feat(core): add change detection tracing service

### PR DESCRIPTION
**Pull request by eneajaho**
This will warn users when they are rendering too much based on the config they provide. Linked to #58540

It works somehow like this: 

```ts
// More than 10 change detections in 1 second triggers a warning.
provideChangeDetectionTracingService({
  msThresholdWindow: 1000,
  cdThreshold: 10,
}),
```

CleanShot 2024-11-26 at 23 22 50

### CODE REVIEW

File: packages/core/src/application/cd-tracing.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import {inject, InjectionToken, Provider} from '../di';
+import {NgZone} from '../zone';
+import {TracingAction, TracingService, TracingSnapshot} from './tracing';
+
+interface CDTracingOptions {
+  msThresholdWindow: number;
+  cdThreshold: number;
+}
+
+const CD_TRACING_OPTIONS = new InjectionToken<CDTracingOptions>('CD_TRACING_OPTIONS');
```
Comment (Yberion) [inline packages/core/src/application/cd-tracing.ts:19]:
```suggestion
const CD_TRACING_OPTIONS = new InjectionToken<CDTracingOptions>(ngDevMode ? 'CD_TRACING_OPTIONS' : '');
```

**Comment by mgechev** [comment]
@eneajaho that's a great idea and it would be a perfect addition to Angular DevTools!

It'd make sense to keep it outside of the framework's API surface so that we keep it focused and avoid the risk of this code ending up in someone's prod bundle.

Would you work with @JeanMeche and @AleksanderBodurri on adding this feature to the profiler?

**Comment by AleksanderBodurri** [comment]
I noticed this is using a service introduced in https://github.com/angular/angular/pull/58771. Seems like the tracing service here performs a similar task to the [existing profiler API](https://github.com/angular/angular/blob/9dbe6fc18be700e89f28a023378f4775f3f7c3fe/packages/core/src/render3/profiler.ts#L36) (see example usage in the comment above), with the primary differences being how the consumer interacts with the API (DI in the case of the tracing service and the ng global in the case of the profiler), and the type of events being captured.

Looping in @crisbeto and @dgp1130 to see how we want to proceed

**Comment by JeanMeche** [comment]
Good Idea, I'll sync with enea & Alex to port this to the devtools.

**Comment by pkozlowski-opensource** [comment]
Closing this PR based on the discussion - the idea is neat, we just need to fold it into the existing tooling instead of creating a new entry point.

**Pull request closed without merging**